src/models/ClassPotentialCxSM.cpp

```diff
@@ -784,50 +733,238 @@ void Class_CxSM::TripleHiggsCouplings()
-  std::size_t posGp = 0, posGm = 0, posG0 = 0;
-  std::size_t posH1 = 0, posH2 = 0, posH3 = 0;
-  const double ZeroThreshold = 1e-5;
+  // CxSM interaction basis
+  // 0   1   2   3      4      5
+  // Gp, Gm, G0, zeta1, zeta2, zeta3
+  int pos_i_G1 = 0, pos_i_G2 = 1, pos_i_G0 = 2;
+  int pos_zeta1 = 3, pos_zeta2 = 4, pos_zeta3 = 5;
 
-  for (std::size_t i = 0; i < NHiggs; i++)
+  // Indices of mass eigenstates for rotation from interaction to mass basis
+  // Only neutral/physical part necessary, as Goldstones are already diagonal
+  pos_Gp = -1, pos_Gm = -1, pos_G0 = -1,
+  pos_h1 = -1, pos_h2 = -1, pos_h3 = -1;
+
+  // First finding the correct rows for the Goldstone bosons
+  for (std::size_t i = 0; i < 3; i++)
   {
-    // the rotation matrix is diagonal besides for the neutral scalars
-    if (std::abs(HiggsRot(i, 0)) > ZeroThreshold)
-      posGp = i;
-    else if (std::abs(HiggsRot(i, 1)) > ZeroThreshold)
-      posGm = i;
-    else if (std::abs(HiggsRot(i, 2)) > ZeroThreshold)
-      posG0 = i;
-
-    // the neutral scalars mix
-    if ((std::abs(HiggsRot(i, 3)) + std::abs(HiggsRot(i, 4)) +
-         std::abs(HiggsRot(i, 5))) > ZeroThreshold)
+    if (std::abs(HiggsRot(i, pos_i_G1))
+        + std::abs(HiggsRot(i, pos_i_G2)) > ZeroThreshold)
     {
-      // use that scalars are sorted by mass
-      if (posH1 == 0)
+      if (pos_Gp == -1)
       {
-        posH1 = i;
+        pos_Gp = i;
       }
       else
       {
-        if (posH2 == 0)
-        {
-          posH2 = i;
-        }
-        else
-        {
-          posH3 = i;
-        }
+        pos_Gm = i;
+      }
+    }
+    else if (std::abs(HiggsRot(i, pos_i_G0)) > ZeroThreshold)
+    {
+      pos_G0 = i;
+    }
+    else
+    {
+      throw std::runtime_error("Error. Non-Goldstone in Goldstone submatrix."
+                               " Particle with negative mass? Check your "
+                               "parameter point.");
+    }
+  }
+
+
+  // Starting from 3 to NHiggs, i.e. fixing the physical Higgs bosons
+  for (std::size_t i = 3; i < NHiggs; i++)
+    // mass base index i corresponds to mass vector sorted in ascending mass
+  {
+    // Neutral submatrix
+    if (std::abs(HiggsRot(i, pos_zeta1)) + std::abs(HiggsRot(i, pos_zeta2))
+        + std::abs(HiggsRot(i, pos_zeta3)) > ZeroThreshold)
+      // use that mh1 < mh2 < mh3
+    {
+      if (pos_h1 == -1) {
+        pos_h1 = i;
+      } else if (pos_h2 == -1) {
+        pos_h2 = i;
+      } else {
+        pos_h3 = i;
+      }
+    }
+  }
+
+  // check if all position indices are set
+  if (pos_h1 == -1 or pos_h2 == -1 or pos_h3 == -1)
+  {
+    throw std::runtime_error("Error. Not all position indices are set.");
+  }
+
+  // check if all other elements of rotation matrix are zero
+  bool zero_element = false;
+  for (std::size_t i = 0; i < NHiggs; i++)
+  {
+    for (std::size_t j = 0; j < NHiggs; j++)
+    {
+      int ii = int(i);
+      int jj = int(j);
+      if (not((jj == pos_Gp    and ii == pos_Gp) or
```

Comment:
This piece of code is a bit confusing, what is it checking for? Can you refactor it a little bit?